In [1]:
%load_ext autoreload
%autoreload 2

In [2]:
import sys
import os
project_root = os.path.abspath(os.path.join(os.getcwd(), '..'))
if project_root not in sys.path:
    sys.path.append(project_root)

In [3]:
from src.text_chunking import VectorStoreBuilder

In [5]:

builder = VectorStoreBuilder()
builder.run_pipeline(sample_size=12000)

2026-06-20 22:13:39,352 - INFO - Loaded cleaned dataset with shape: (464010, 18)
2026-06-20 22:13:39,730 - INFO - Created stratified sample of 11998 complaints
2026-06-20 22:13:39,740 - INFO - Distribution:
shape: (4, 2)
┌──────────────────┬──────┐
│ Product_Category ┆ len  │
│ ---              ┆ ---  │
│ str              ┆ u32  │
╞══════════════════╪══════╡
│ Credit Card      ┆ 5097 │
│ Personal Loan    ┆ 721  │
│ Savings Account  ┆ 3628 │
│ Money Transfer   ┆ 2552 │
└──────────────────┴──────┘
2026-06-20 22:13:41,302 - INFO - Created 34605 text chunks from 11998 complaints
2026-06-20 22:13:41,304 - INFO - Loading embedding model: sentence-transformers/all-MiniLM-L6-v2
2026-06-20 22:13:41,315 - INFO - No device provided, using cpu
2026-06-20 22:13:43,772 - INFO - HTTP Request: GET https://huggingface.co/api/agent-harnesses "HTTP/1.1 200 OK"
2026-06-20 22:13:44,419 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/modules.json "HTTP/

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

c:\Users\Hello\Desktop\kaim\rag-complaint-chatbot\venv\Lib\site-packages\huggingface_hub\file_download.py:137: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\Hello\.cache\huggingface\hub\models--sentence-transformers--all-MiniLM-L6-v2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)
2026-06-20 22:13:45,438 - INFO - HTTP Request: HEAD https://huggingface.co/

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

2026-06-20 22:13:46,032 - INFO - Loading SentenceTransformer model from sentence-transformers/all-MiniLM-L6-v2.
2026-06-20 22:13:46,375 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/config_sentence_transformers.json "HTTP/1.1 307 Temporary Redirect"
2026-06-20 22:13:46,771 - INFO - HTTP Request: HEAD https://huggingface.co/api/resolve-cache/models/sentence-transformers/all-MiniLM-L6-v2/1110a243fdf4706b3f48f1d95db1a4f5529b4d41/config_sentence_transformers.json "HTTP/1.1 200 OK"
2026-06-20 22:13:47,125 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/README.md "HTTP/1.1 307 Temporary Redirect"
2026-06-20 22:13:47,359 - INFO - HTTP Request: HEAD https://huggingface.co/api/resolve-cache/models/sentence-transformers/all-MiniLM-L6-v2/1110a243fdf4706b3f48f1d95db1a4f5529b4d41/README.md "HTTP/1.1 200 OK"
2026-06-20 22:13:47,604 - INFO - HTTP Request: GET https://huggingface.co/api/reso

README.md:   0%|          | 0.00/10.5k [00:00<?, ?B/s]

2026-06-20 22:13:48,049 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/modules.json "HTTP/1.1 307 Temporary Redirect"
2026-06-20 22:13:48,286 - INFO - HTTP Request: HEAD https://huggingface.co/api/resolve-cache/models/sentence-transformers/all-MiniLM-L6-v2/1110a243fdf4706b3f48f1d95db1a4f5529b4d41/modules.json "HTTP/1.1 200 OK"
2026-06-20 22:13:48,653 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/sentence_bert_config.json "HTTP/1.1 307 Temporary Redirect"
2026-06-20 22:13:48,885 - INFO - HTTP Request: HEAD https://huggingface.co/api/resolve-cache/models/sentence-transformers/all-MiniLM-L6-v2/1110a243fdf4706b3f48f1d95db1a4f5529b4d41/sentence_bert_config.json "HTTP/1.1 200 OK"
2026-06-20 22:13:50,039 - INFO - HTTP Request: GET https://huggingface.co/api/resolve-cache/models/sentence-transformers/all-MiniLM-L6-v2/1110a243fdf4706b3f48f1d95db1a4f5529b4d41/sentence_bert_config.json

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

2026-06-20 22:13:50,571 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/adapter_config.json "HTTP/1.1 404 Not Found"
2026-06-20 22:13:51,039 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/config.json "HTTP/1.1 307 Temporary Redirect"
2026-06-20 22:13:51,771 - INFO - HTTP Request: HEAD https://huggingface.co/api/resolve-cache/models/sentence-transformers/all-MiniLM-L6-v2/1110a243fdf4706b3f48f1d95db1a4f5529b4d41/config.json "HTTP/1.1 200 OK"
2026-06-20 22:13:52,025 - INFO - HTTP Request: GET https://huggingface.co/api/resolve-cache/models/sentence-transformers/all-MiniLM-L6-v2/1110a243fdf4706b3f48f1d95db1a4f5529b4d41/config.json "HTTP/1.1 200 OK"


config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

2026-06-20 22:13:52,893 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/model.safetensors "HTTP/1.1 302 Found"


model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

2026-06-20 22:52:27,464 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/processor_config.json "HTTP/1.1 404 Not Found"
2026-06-20 22:52:27,807 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/preprocessor_config.json "HTTP/1.1 404 Not Found"
2026-06-20 22:52:28,152 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/video_preprocessor_config.json "HTTP/1.1 404 Not Found"
2026-06-20 22:52:28,499 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/preprocessor_config.json "HTTP/1.1 404 Not Found"
2026-06-20 22:52:28,836 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/tokenizer_config.json "HTTP/1.1 307 Temporary Redirect"
2026-06-20 22:52:29,574 - INFO - HTTP Request: HEAD https://huggingface.co/api/resolve-cache/models/senten

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

2026-06-20 22:52:30,242 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/config.json "HTTP/1.1 307 Temporary Redirect"
2026-06-20 22:52:30,475 - INFO - HTTP Request: HEAD https://huggingface.co/api/resolve-cache/models/sentence-transformers/all-MiniLM-L6-v2/1110a243fdf4706b3f48f1d95db1a4f5529b4d41/config.json "HTTP/1.1 200 OK"
2026-06-20 22:52:32,028 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/config.json "HTTP/1.1 307 Temporary Redirect"
2026-06-20 22:52:32,255 - INFO - HTTP Request: HEAD https://huggingface.co/api/resolve-cache/models/sentence-transformers/all-MiniLM-L6-v2/1110a243fdf4706b3f48f1d95db1a4f5529b4d41/config.json "HTTP/1.1 200 OK"
2026-06-20 22:52:33,333 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/tokenizer_config.json "HTTP/1.1 307 Temporary Redirect"
2026-06-20 22:52:34,045 - INFO - HTTP Request: HEAD

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

2026-06-20 22:52:37,107 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/tokenizer.json "HTTP/1.1 307 Temporary Redirect"
2026-06-20 22:52:37,336 - INFO - HTTP Request: HEAD https://huggingface.co/api/resolve-cache/models/sentence-transformers/all-MiniLM-L6-v2/1110a243fdf4706b3f48f1d95db1a4f5529b4d41/tokenizer.json "HTTP/1.1 200 OK"
2026-06-20 22:52:37,813 - INFO - HTTP Request: GET https://huggingface.co/api/resolve-cache/models/sentence-transformers/all-MiniLM-L6-v2/1110a243fdf4706b3f48f1d95db1a4f5529b4d41/tokenizer.json "HTTP/1.1 200 OK"


tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

2026-06-20 22:52:39,113 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/added_tokens.json "HTTP/1.1 404 Not Found"
2026-06-20 22:52:39,744 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/special_tokens_map.json "HTTP/1.1 307 Temporary Redirect"
2026-06-20 22:52:39,972 - INFO - HTTP Request: HEAD https://huggingface.co/api/resolve-cache/models/sentence-transformers/all-MiniLM-L6-v2/1110a243fdf4706b3f48f1d95db1a4f5529b4d41/special_tokens_map.json "HTTP/1.1 200 OK"
2026-06-20 22:52:40,206 - INFO - HTTP Request: GET https://huggingface.co/api/resolve-cache/models/sentence-transformers/all-MiniLM-L6-v2/1110a243fdf4706b3f48f1d95db1a4f5529b4d41/special_tokens_map.json "HTTP/1.1 200 OK"


special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

2026-06-20 22:52:40,762 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/chat_template.jinja "HTTP/1.1 404 Not Found"
2026-06-20 22:52:41,304 - INFO - HTTP Request: HEAD https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/resolve/main/1_Pooling/config.json "HTTP/1.1 307 Temporary Redirect"
2026-06-20 22:52:41,533 - INFO - HTTP Request: HEAD https://huggingface.co/api/resolve-cache/models/sentence-transformers/all-MiniLM-L6-v2/1110a243fdf4706b3f48f1d95db1a4f5529b4d41/1_Pooling%2Fconfig.json "HTTP/1.1 200 OK"
2026-06-20 22:52:41,766 - INFO - HTTP Request: GET https://huggingface.co/api/resolve-cache/models/sentence-transformers/all-MiniLM-L6-v2/1110a243fdf4706b3f48f1d95db1a4f5529b4d41/1_Pooling%2Fconfig.json "HTTP/1.1 200 OK"


config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

2026-06-20 22:52:42,386 - INFO - HTTP Request: GET https://huggingface.co/api/models/sentence-transformers/all-MiniLM-L6-v2 "HTTP/1.1 200 OK"
2026-06-20 22:52:42,402 - INFO - Generating embeddings... (this may take a few minutes)


Batches:   0%|          | 0/271 [00:00<?, ?it/s]

2026-06-20 23:08:18,211 - INFO - Generated embeddings with shape: (34605, 384)
2026-06-20 23:08:18,478 - INFO - Vector store successfully saved to ../vector_store/
2026-06-20 23:08:18,493 - INFO - ✅ Vector Store Pipeline Completed Successfully!




#### **Task 2 Technical Report**
* **Sampling Strategy:** To distill the refined dataset, a *Stratified Random Sampling* framework was leveraged to target exactly 12,000 narratives. A simple random sample would introduce severe category risk due to the underlying class imbalances established in our EDA. Stratification ensures that minor classes (like Personal Loans) maintain their proportional weight (~6% of the vector index space), preventing the RAG system from developing a retrieval bias towards high-volume classes like Credit Cards.

* **Chunking Approach:** The pipeline utilizes LangChain's `RecursiveCharacterTextSplitter` configured to a strict window of `512` characters with a `50`-character recursive overlap. This specific size splits long, right-skewed narrative complaints into semantic windows containing approximately 2–3 sentences. This prevents the "dilution" of fine-grained details during vector pooling. The 50-character buffer handles overlapping boundaries, ensuring that critical data points (such as dates, account names, or specific violation keywords) split across boundaries are not lost to retrieval. The pipeline processed 12,000 sampled rows into exactly 34,605 clean text nodes.

* **Embedding Model Choice:** The open-source `all-MiniLM-L6-v2` transformer model was selected. This model compresses text chunks into compact 384-dimensional vectors. It was chosen because it represents an outstanding balance between local computational throughput and semantic accuracy. It requires minimal hardware memory footprint while maintaining high score metrics for capturing the natural, messy vocabulary typical of consumer financial disputes.

* **Indexing Architecture:** Vector search is managed using native `FAISS` running an Inner Product space (`IndexFlatIP`). Because all generated vectors are pre-normalized, calculating the inner product functions as a high-speed Cosine Similarity retrieval mechanism. Traceability is guaranteed by serializing an isolated mapping index (`metadata.pkl`), which anchors every generated chunk directly to its parent `Complaint ID` and `Product_Category` for accurate source-attribution during generation.
